packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD2.hpp

```diff
@@ -0,0 +1,1323 @@
+	      femv_colors(curr_lid) = 0;
+	      recoloring_size++;
+	      break;
+	    } else if (vid_d1_degree < curr_degree && recolor_degrees){
+	      femv_colors(vid_d1) = 0;
+	      recoloring_size++;
+	    } else if(rand_dev(curr_lid) < rand_dev(vid_d1)){
+	      femv_colors(curr_lid) = 0;
+	      recoloring_size++;
+	      break;
+	    } else if(rand_dev(vid_d1) < rand_dev(curr_lid)){
+	      femv_colors(vid_d1) = 0;
+	      recoloring_size++;
+	    } else{
+	      if(gid_dev(curr_lid) >= gid_dev(vid_d1)){
+	        femv_colors(curr_lid) = 0;
+		recoloring_size++;
+		break;
+	      } else {
+	        femv_colors(vid_d1) = 0;
+	        recoloring_size++;
+	      }
+	    }
+	  }
+	  size_t d2_adj_begin = 0;
+	  size_t d2_adj_end   = 0;
+	  d2_adj_begin = dist_offsets_dev(vid_d1);
+	  d2_adj_end   = dist_offsets_dev(vid_d1+1);
+	  bool found = false;
+	  for(size_t vid_d2_adj = d2_adj_begin; vid_d2_adj < d2_adj_end; vid_d2_adj++){
+	    const size_t vid_d2 = dist_adjs_dev(vid_d2_adj);
+	    size_t vid_d2_degree = 0;
+	    if(vid_d2 < n_local){
+	      vid_d2_degree = dist_offsets_dev(vid_d2+1) - dist_offsets_dev(vid_d2);
+	    } else {
+	      vid_d2_degree = ghost_degrees_dev(vid_d2-n_local);
+	    }
+	    if(curr_lid != vid_d2 && femv_colors(vid_d2) == curr_color){
+	      if(curr_degree < vid_d2_degree && recolor_degrees){
+                found = true;
+		femv_colors(curr_lid) = 0;
+		recoloring_size++;
+		break;
+	      } else if(vid_d2_degree < curr_degree && recolor_degrees){
+		femv_colors(vid_d2) = 0;
+		recoloring_size++;
+	      } else if(rand_dev(curr_lid) < rand_dev(vid_d2)){
+	        found = true;
+	        femv_colors(curr_lid) = 0;
+		recoloring_size++;
+	        break;
+	      } else if(rand_dev(vid_d2) < rand_dev(curr_lid)){
+	        femv_colors(vid_d2) = 0;
+	        recoloring_size++;
+	      } else {
+	        if(gid_dev(curr_lid) >= gid_dev(vid_d2)){
+	          found = true;
+	          femv_colors(curr_lid) = 0;
+		  recoloring_size++;
+	          break;
+	        } else {
+	          femv_colors(vid_d2) = 0;
+		  recoloring_size++;
+	        }
+ 	      }
+	    }
+	  }
+       	  if(found) break;
+        }
+      },recoloringSize(0));
+      Kokkos::fence();
+      verts_to_send_size(0) = 0;
+      verts_to_recolor_size(0) = 0;
+      Kokkos::parallel_for(femv_colors.size(), KOKKOS_LAMBDA(const uint64_t& i){
+        if(femv_colors(i) == 0){
+	  if(i < n_local){
+	    verts_to_send_atomic(verts_to_send_size_atomic(0)++) = i;
+	  }
+	  verts_to_recolor_atomic(verts_to_recolor_size_atomic(0)++) = i;
+	}
+      });
+      Kokkos::fence();
+      if(comm->getSize() > 1){
+        conflict_detection = timer() - temp;
+        comp_time += conflict_detection;
+      }
+      //ensure that the parallel_for finishes before continuing
+      if(verbose) std::cout<<comm->getRank()<<": done checking initial conflicts\n";
+     
+      //all conflicts detected!
+
+      //variables for statistics
+      double totalPerRound[100];
```

Comment:
Add comments:  what are these arrays storing?  Are there guards to prevent their size 100 from being exceeded?